brandon_tan/partner_selector.py

```diff
@@ -0,0 +1,483 @@
+        :type method: str
+        :param estimator: The extended method has 3 estimators. This parameter identifies which estimators to be used. Defaults to 1.
+        :type estimator: int, optional
+        :raises Exception: In the event an invalid method is chosen, error will be raised.
+        :return: The best partners based on user chosen method will be returned in a list format.
+        :rtype: list
+        """
+
+        method = method.lower()
+        if method == 'traditional':
+            best_Q = self._traditional_method()
+
+        elif method == 'extended':
+            best_Q = self._extended_method(estimator)
+
+        elif method == 'geometric':
+            best_Q = self._geometric_method()
+
+        else:
+            raise Exception("Unknown method. Either 'traditional' , 'extended' or 'geometric'.")
+
+        return best_Q
+
+    def _traditional_method(self) -> list:
+        """
+        The function loops through all possible partner combinations and calculates the sum of all pairwise correlations.
+        The largest sum of pairwise correlation will be returned.
+
+        :return: The best partners based on sum of pairwise correlation will be returned.
+        :rtype: list
+        """
+
+        highest_corr = 0
+        partner_data = {}
+        for i in tqdm(range(len(self.all_possible_partners))):
+
+            # Converting to list as itertools combination generates an itertools object.
+            partners = list(self.all_possible_partners[i])
+
+            # When generating the partner combinations, we excluded the target ticker
+            # and hence append it back here.
+            # Append (instead of insert) was chosen due to O(1) complexity.
+            partners.append(self.target_ticker)
+
+            # Finding the sum of the right upper triangle of the correlation matrix.
+            sum_of_pairwise_corr = self._upper_tri_sum(self.ranked_returns_data_corr.loc[partners, partners].values)
+
+            if sum_of_pairwise_corr > highest_corr:
+                highest_corr = sum_of_pairwise_corr
+                best_Q = partners
+
+            partner_data[i] = {
+                'Grouping': partners,
+                'Spearman Bivariate Correlation Sum': sum_of_pairwise_corr
+            }
+
+        self.partner_data = partner_data
+        self.best_Q = best_Q
+        return best_Q
+
+    def _upper_tri_sum(self,corr_matrix):
+        """
+        Helper function for traditional method. Given that
+        we want the sum of pairwise correlation, it is essentially the
+        upper triangular of the correlation matrix, without the
+        diagonal. We use a mask to identify that triangle and sum it.
+
+        :param corr_matrix: Spearman correlation matrix.
+        :type corr_matrix: np.ndarray
+        :return: Sum of upper triangular, less the diagonal.
+        :rtype: float
+        """
+        m = corr_matrix.shape[0]
```

Comment:
These variable names are bad. Rather use full descriptions. This won't pass a linter.

Reply:
Noted on this. If I may ask, what would be a better naming for "m"? 

My understanding is that in linear algebra, matrices are denoted as m by n. Hence "m" might be a good naming.